pypeit/scripts/identify.py

```diff
@@ -86,55 +96,240 @@ def main(args):
+                # Load the calibration frame (if it exists and is desired).  Bad-pixel mask
+                # set to any flagged pixel in Arc.
+                wv_calib_slit = None
+                if wv_calib is not None:
+                    if not np.any(wv_calib.wv_fits):
+                        wv_calib_slit = None
+                    else:
+                        if not wv_calib.wv_fits[slit_val]['pypeitfit']:
+                            wv_calib_slit = None
+                        else:
+                            wv_calib_slit = wv_calib
 
-        # Load the calibration frame (if it exists and is desired).  Bad-pixel mask
-        # set to any flagged pixel in Arc.
-        wavecal = BuildWaveCalib(msarc, slits, spec, par, lamps, det=args.det,
-                                 msbpm=msarc.select_flag())
-        arccen, arc_maskslit = wavecal.extract_arcs(slitIDs=[args.slit])
-
-        # Launch the identify window
-        # TODO -- REMOVE THIS HACK
-        try:
-            nonlinear_counts = msarc.detector.nonlinear_counts()
-        except AttributeError:
-            nonlinear_counts = None
-        arcfitter = Identify.initialise(arccen, lamps, slits, slit=int(args.slit), par=par,
-                                        wv_calib_all=wv_calib, wavelim=[args.wmin, args.wmax],
-                                        nonlinear_counts=nonlinear_counts,
-                                        pxtoler=args.pixtol, test=args.test, 
-                                        fwhm=args.fwhm,
-                                        sigdetect=args.sigdetect,
-                                        specname=spec.name,
-                                        y_log=not args.linear,
-                                        rescale_resid=args.rescale_resid)
-
-        # If testing, return now
-        if args.test:
-            return arcfitter
-        final_fit = arcfitter.get_results()
-
-        # Build here to avoid circular import
-        #  Note:  This needs to be duplicated in test_scripts.py
-        # Wavecalib (wanted when dealing with multiple detectors, eg. GMOS)
-        if 'WaveFit' in arcfitter._fitdict.keys():
-            waveCalib = WaveCalib(nslits=1, wv_fits=np.atleast_1d(arcfitter._fitdict['WaveFit']),
-                                  arc_spectra=np.atleast_2d(arcfitter.specdata).T,
-                                  spat_ids=np.atleast_1d(int(arcfitter._spatid)),
-                                  PYP_SPEC=msarc.PYP_SPEC, lamps=','.join(lamps))
+                arccen, arc_maskslit = wavecal.extract_arcs(slitIDs=[slit_val])
+
+                # Launch the identify window
+                # TODO -- REMOVE THIS HACK
+                try:
+                    nonlinear_counts = msarc.detector.nonlinear_counts()
+                except AttributeError:
+                    nonlinear_counts = None
+                arcfitter = Identify.initialise(arccen, lamps, slits, slit=int(slit_val), par=par,
```

Comment:
I am really not a fan of this `if args.multi` statement. I think this needs to be refactored to limit the amount of duplicated code. By making these two "streams" of multi=True or multi=False, is going to be prone to errors and harder to maintain long term. I'd encourage you to refactor this so that there's only one call to `initialise`, and one call to `store_solution`.

Also, I wonder if the code would benefit from breaking off some of the bits into small functions?

Reply:
I have consolidated the two cases.